### DISCUSSION TITLE
Definition of density for large strain problems (UpdatedLagrangian formulation)

### DISCUSSION THREAD

**Discussion by saturn00000** [idaholab/moose#22467] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE Team,

as you can see in the attached picture. if we try to solve heat transfer problems at **large deformations** (_"strain = FINITE", UpdatedLagrangian formulation), do we need to modify the **density \rho** in the codes?
e.g.
` _rho[_qp] ->  _rho[_qp]*det(F) `
or
`_rho[_qp] ->  _rho[_qp]*J `

[Modules/TensorMechanics/Master]
  [all]
    # This block adds all of the proper Kernels, strain calculators, and Variables
    # for TensorMechanics in the correct coordinate system (autodetected)
    add_variables = true
    strain = FINITE
    eigenstrain_names = eigenstrain
    use_automatic_differentiation = true
    generate_output = 'vonmises_stress elastic_strain_xx elastic_strain_yy strain_xx strain_yy'
  []
[]

(ref: moose\tutorials\darcy_thermo_mech\step09_mechanics\problems)

![grafik](https://user-images.githubusercontent.com/107309988/197307412-269137ad-5432-4250-85bd-3c0e2ace1fa5.png)

**Comment by hugary1995** [comment, marked as answer]
You have three choices:
1. In the reference configuration, implement `_test[_i][_qp] * _rho[_qp] * _e_dot[_qp]  * _F[_qp].det()`.
2. In the reference configuration, implement `_test[_i][_qp] * _rho_0[_qp] * _e_dot[_qp]`. This is the right half of your picture.
3. In the current (deformed) configuration, implement `_test[_i][_qp] * _rho[_qp] * _e_dot[_qp]`. This is the left half of your picture.

To integrate on the undisplaced mesh, you set `use_displaced_mesh = false`. To integrate on the displaced mesh, you set `use_displaced_mesh = true`. In cases 1 and 3, you will specify the density of the material in the deformed configuration, e.g. \rho. In case 2, you will specify the density of the material in the reference configuration, e.g. \rho_0.

As you can see, cases 2 and 3 have the same functional form. So that's what most moose objects offer, as you can easily switch between them using `use_displaced_mesh`. Case 1 is still useful if you don't want to keep track of the displaced mesh when you only have the measurement of the density in the deformed configuration.

**Reply by hugary1995** [reply]
By "Lagrangian formulation", are you referring to the new system in tensor mechanics, e.g. `new_system = true` in the action? We always do solid mechanics using a Lagrangian description, at least in the tensor_mechanics module. Though sometimes we distinguish between total Lagrangian and updated Lagrangian.

**Reply by saturn00000** [reply]
Dear @hugary1995 

for **UpdatedLagrangian formulation** in Moose, you may see moose\tutorials\darcy_thermo_mech\step09_mechanics\problems\step9.i

In step9.i,  the density parameters are defined in the reference configuration (i.e.  \rho_{0}). These density paramters in step9.i are \rho_{0}, not \rho_{t}.

As you know, if we use **UpdatedLagrangian formulation** in MOOSE, we need to use \rho_{t}, not \rho_{0}...

**Reply by hugary1995** [reply]
That probably is an oversight. Is this from a tutorial you were trying to follow? If so, can you give me the link to that tutorial? I'll create an issue if I can confirm this.

**Reply by saturn00000** [reply]
@hugary1995 the link to the tutorial: [https://github.com/idaholab/moose/blob/next/tutorials/darcy_thermo_mech/step09_mechanics/problems/step9.i]

We may note that:
1.)  in your "raccoon" modules,  total Lagrangian is used, therefore, \rho_{0} is okay, as we do everything in the reference configuration

2.) But for  [Modules/TensorMechanics/Master]
[all]
strain = FINITE 
...
[]

UpdatedLagrangian formulation is used, thus everything in the deformed configuration.  \rho_{0} is NOT okay, **OR**?

**Reply by hugary1995** [reply]
> UpdatedLagrangian formulation is used, thus everything in the deformed configuration.

This statement is not entirely correct. We still have the freedom to perform the integration on whichever configuration we like. 

> In step9.i, the density parameters are defined in the reference configuration (i.e. \rho_{0}).

I'm not sure I follow. Why do you think the density values in that input file are \rho_0?

**Reply by saturn00000** [reply]
> > UpdatedLagrangian formulation is used, thus everything in the deformed configuration.
> 
> This statement is not entirely correct. We still have the freedom to perform the integration on whichever configuration we like.
> 
is  the updatedLagrangian formulation  used in **step9.i**?  or NOT ?

 [Modules/TensorMechanics/Master]
[all]
strain = FINITE
...
[]
If yes,  this statement is entirely correct in this case.
> > In step9.i, the density parameters are defined in the reference configuration (i.e. \rho_{0}).
> 
> I'm not sure I follow. Why do you think the density values in that input file are \rho_0?

can you find line 126 in **step9.i**:   `density_file = data/water_density.csv`
here we define the density ... ...

how to define the density value **in the input file** for solving the updatedLagrangian formulation problems in MOOSE?

Please note that the density \rho for large deformation problems is NOT constant in case of  the updatedLagrangian formulation...

**Reply by hugary1995** [reply]
The tutorial is doing the right thing (case 2 in my original answer).

**Reply by saturn00000** [reply]
The **updatedLagrangian** formulation is used in step9.i (the tutorial you mean)...
If the **updatedLagrangian** formulation is used in this case, why is the density \rho **constant**?
Please see
![grafik](https://user-images.githubusercontent.com/107309988/197615581-8e172dd3-5e4b-4d10-9cdc-afd61b609739.png)

**Reply by saturn00000** [reply]
Hello @hugary1995 

step9.i may need some modifications for , e.g.
```
...
    density /= detF;
  }

  _density[_qp] = density;
```
  **density /= detF;**

Ref : **Density.C** and  **power_law_creep_restart1.i** in MOOSE

**Reply by hugary1995** [reply]
The implementation in Density.C is also correct. It defines \rho in the deformed configuration.

Again, in which configuration you choose to integrate the kernel is irrelevant to the objective frame.

**Reply by saturn00000** [reply]
Hello @hugary1995

in step9.i, the displacement field is implemented in deformed configurations, other fields are implemented in undeformed configurations, right?

**Reply by hugary1995** [reply]
Before discretization, the function space for the reference configuration C^0(\Omega) \cap H^1_0(\Omega) and the deformed configuration C^0(\Omega_t) \cap H^1_0(\Omega_t) are indeed different, as long as the deformation map is not the identity map. However, once you discretize them with Lagrange polynomials, the same discrete vector can be used to express both the displacement field in the reference configuration as well as in the deformed configuration, as the polynomials are orthonormal.
